neps/nep-0621.md

```diff
@@ -0,0 +1,491 @@
+    /// - Since transactions on NEAR are non-atomic, the vault contract should follow one of these approaches:
+    ///   1. Emit a `VaultWithdraw` event when the fee is deducted, and if the withdrawal later fails, emit a `VaultDeposit` event
+    ///      to revert the funds; **or**
+    ///   2. Emit a `VaultWithdraw` event **only** if the withdrawal succeeds.
+    /// - In the reference implementation, the contract emits an `FtBurn` event when the user calls the `withdraw` method.
+    ///   If the withdrawal is successful, a `VaultWithdraw` event is emitted in the `resolve_withdraw` callback.
+    ///   If the withdrawal fails, an `FtMint` event is emitted instead to restore the user’s balance.
+    ///
+    /// See also: [`Self::preview_withdraw`].
+    fn withdraw(&mut self, assets: U128, receiver_id: Option<AccountId>) -> PromiseOrValue<U128>;
+
+    /// Simulates withdrawing exactly `assets` worth of tokens without executing.
+    ///
+    /// Differs from [`Self::convert_to_shares`] by factoring in:
+    /// - The caller's current share balance.
+    /// - Vault withdrawal limits.
+    /// - Applicable fees or penalties.
+    ///
+    /// # Panics / Fails
+    /// - The `preview_withdraw` method must never panic.
+    ///   It should return the number of shares required to withdraw the given amount of assets,
+    ///   or `0` if withdrawals are not currently allowed.
+    ///
+    /// Useful for frontends to preview required shares for a given withdrawal.
+    fn preview_withdraw(&self, assets: U128) -> U128;
+}
+```
+
+### Events
+
+```rust
+/// Event emitted when a deposit is received by the vault.
+///
+/// This follows the proposed NEP vault standard, referencing the ERC-4626 pattern.
+/// Upon receiving assets, the vault mints and issues shares to the `owner_id`.
+pub struct VaultDeposit {
+    /// The account that sends the deposit (payer of the assets).
+    pub sender_id: AccountId,
+
+    /// The account that receives the minted shares.
+    pub owner_id: AccountId,
+
+    /// Amount of underlying assets deposited into the vault.
+    pub assets: U128,
+
+    /// Amount of shares minted and issued to `owner_id`.
+    pub shares: U128,
+
+    /// Optional memo provided by the sender for off-chain use.
+    pub memo: Option<String>,
+}
+
+/// Event emitted when shares are redeemed from the vault.
+///
+/// Upon redemption, the vault burns the shares from `owner_id`
+/// and transfers the equivalent assets to `receiver_id`.
+pub struct VaultWithdraw {
+    /// The account that owns the shares being redeemed (burned).
+    pub owner_id: AccountId,
+
+    /// The account receiving the underlying assets.
+    pub receiver_id: AccountId,
+
+    /// Amount of shares redeemed (burned from the vault).
+    pub shares: U128,
+
+    /// Amount of underlying assets withdrawn from the vault.
+    pub assets: U128,
+
+    /// Optional memo provided by the redeemer for off-chain use.
+    pub memo: Option<String>,
+}
+```
+
+Deposit and Mint
```

Comment:
shouldn't this section be placed inside the "Contract Interface" section and not inside "Events" section?